com/bl/ttt/tictactoe.py

```diff
@@ -49,15 +50,23 @@ def get_previous_turn_player(self):
         else:
             return self.player1
 
-    def get_player_move_log(self):
-        return self.player_move_log
+    def get_player_move_log(self, player_id='both'):
+        if player_id == 'both':
+            x_move_log = list(self.player_move_log.get('X'))
+            o_move_log = list(self.player_move_log.get('O'))
+            players_move_tuples = list(itertools.zip_longest(x_move_log, o_move_log))
+            players_move_list = list(itertools.chain(*players_move_tuples))
+            if len(players_move_list) == 0:
+                return []
+            if players_move_list[-1] is None:
+                players_move_list.pop()
```

Comment:
How about filtering out None with a lit comprehension like this : 

`return [e for e in players_move_list if(e is not None ) ]`

or

`return list( filter( lambda x: x is not None, players_move_list ) )`

Reply:
Try to use the language where ever possible instead of writing your own code.

The code behind the language has been tested billions of times whereas any code we write is vulnerable.

Reply:
There are usually at least two ways to deliver a solution - one that delegates to the language libraries as opposed to writing our own code is far superior for several reasons : 
1. Language libs have been written and tested by devs who are way better that us.
2. Language libs have been run (and therefore tested) billions of times.
3. Lang libs in future releases may have optimised the algo for doing xyz() which we get for free without a single change in our code.
4. The underlying implementation of any language lib is in C which will run much faster than our compiled python code.
5. We have less code to write and test.
6. Our code is easier to read and reason about because there is less of it and the concepts are familiar to other devs who know the language.
  e.g filter above is easier to understand than the if statement you wrote.

Reply:
Roger all those points: thanks for the detailed clarity... I'm off to document it.